### DISCUSSION TITLE
Question about Subtree Extraction during Discriminator Finalization in TTT

### DISCUSSION THREAD

**Discussion by raufum** [LearnLib/learnlib#100] [Q&A]
I am currently reading the dissertation from Malte Isberner who has invented the TTT algorithm

There is a part of the algorithm that I dont understand during the subtree extraction in the discriminator finalization:

the algorithm (which is also implemented in learnlib):

![52](https://github.com/LearnLib/learnlib/assets/103262544/649ada38-6cdf-438c-aecb-e5b2c677d266)
![53](https://github.com/LearnLib/learnlib/assets/103262544/40ac7d08-bb1b-4c20-a6ac-23165f757e5d)
![54](https://github.com/LearnLib/learnlib/assets/103262544/44b747f4-f0c1-4bd4-a833-232531288c7d)

and the part that I dont understand:

![subtreeextraction](https://github.com/LearnLib/learnlib/assets/103262544/d16a968a-f96a-4205-9b04-5f5456a75b2f)

can you maybe give an automata where the respective part of the code is visited? I.e. line 12 or 27 in 5.3

**Comment by mtf90** [comment]
Unfortunately, I am not that farmiliar with the internals of the algorithm, but from looking at the code, I believe the `extract` function corresponds to the [`extractSubtree`](https://github.com/LearnLib/learnlib/blob/develop/algorithms/active/ttt/src/main/java/de/learnlib/algorithm/ttt/base/AbstractTTTLearner.java#L663) method. A quick run of some of the integration-tests shows that the [TTTLearnerMealyIT](https://github.com/LearnLib/learnlib/blob/develop/algorithms/active/ttt/src/test/java/de/learnlib/algorithm/ttt/mealy/it/TTTLearnerMealyIT.java) seems to hit some of the `createNewState` statements. However, the code corresponding to lines 11-13 seems to be [uncovered](https://coveralls.io/builds/63609615/source?filename=algorithms%2Factive%2Fttt%2Fsrc%2Fmain%2Fjava%2Fde%2Flearnlib%2Falgorithm%2Fttt%2Fbase%2FAbstractTTTLearner.java#L684) even by our whole integration-test suite.

However, that might be because we typically use short counterexamples in our tests and this effect may materialize only with counterexamples that actually contain redundancy. Maybe you can try to experiment yourself with some (possibly randomly generated) systems and different counterexample algorithms.

Maybe @misberner finds some time to answer himself.
